src/prefetch.cpp

```diff
@@ -0,0 +1,325 @@
+        LOG_INFO("Prefetch: run with mode `, trace file is `", m_mode, trace_file_path);
+
+        // Open trace file
+        if (m_mode != Mode::Disabled) {
+            int flags = m_mode == Mode::Record ? O_WRONLY : O_RDONLY;
+            m_trace_file = FileSystem::open_localfile_adaptor(trace_file_path.c_str(), flags, 0666, 2);
+        }
+
+        // Loop detect lock file if going to record
+        if (m_mode == Mode::Record) {
+            int lock_fd = open(m_lock_file_path.c_str(), O_RDWR | O_CREAT | O_TRUNC | O_EXCL, 0666);
+            close(lock_fd);
+            photon::thread_create11(&PrefetcherImpl::detect_lock, this);
+        }
+
+        // Reload if going to replay
+        if (m_mode == Mode::Replay) {
+            m_replay_threads.resize(REPLAY_CONCURRENCY);
+            reload(file_size);
+        }
+    }
+
+    ~PrefetcherImpl() {
+        if (m_mode == Mode::Record && !m_record_stopped) {
+            dump();
+        } else if (m_mode == Mode::Replay) {
+            m_replay_stopped = true;
+            for (auto th : m_replay_threads) {
+                photon::thread_shutdown((photon::thread*) th);
+                photon::thread_join(th);
+            }
+        }
+        if (m_trace_file != nullptr) {
+            m_trace_file->close();
+            m_trace_file = nullptr;
+        }
+    }
+
+    IFile* new_prefetch_file(IFile* src_file, uint32_t layer_index) override {
+        return new PrefetchFile(src_file, layer_index, this);
+    }
+
+    void record(TraceOp op, uint32_t layer_index, size_t count, off_t offset) override {
+        if (m_record_stopped) {
+            return;
+        }
+        TraceFormat trace = {op, layer_index, count, offset};
+        m_record_array.push_back(trace);
+    }
+
+    void replay() override {
+        if (m_mode != Mode::Replay) {
+            return;
+        }
+        if (m_replay_queue.empty() || m_src_files.empty()) {
+            return;
+        }
+        LOG_INFO("Prefetch: Replay ` records from ` layers", m_replay_queue.size(), m_src_files.size());
+        for (int i = 0; i < REPLAY_CONCURRENCY; ++i) {
+            auto th = photon::thread_create11(&PrefetcherImpl::replay_worker_thread, this);
+            m_replay_threads[i] = photon::thread_enable_join(th);
+        }
+    }
+
+    int replay_worker_thread() {
+        static char buf[MAX_IO_SIZE];       // multi threads reuse one buffer
+        while (!m_replay_queue.empty() && !m_replay_stopped) {
+            auto trace = m_replay_queue.front();
+            m_replay_queue.pop();
+            auto iter = m_src_files.find(trace.layer_index);
+            if (iter == m_src_files.end()) {
+                continue;
+            }
+            auto src_file = iter->second;
+            if (trace.op == PrefetcherImpl::TraceOp::READ) {
+                ssize_t n_read = src_file->pread(buf, trace.count, trace.offset);
+                if (n_read != (ssize_t) trace.count) {
```

Comment:
here checked the result (count) of prefetch read, but traced `count` based on only `pread` request (in `PrefetchFile::pread`) , not result.